### PR TITLE
Simplify regression template to minimal working example

### PR DESCRIPTION
**Pull request by bbarnes52**
## What changes are proposed in this pull request?

This PR deletes code in the regression template that is not strictly required to demonstrate a working example. Note that some choices are debatable, and some anti-patterns are introduced, such as using features like hour_of_day directly, e.g. without encoding. Tests are removed (these were largely performing type assertions  / null checks anyways) as was documentation, which is expected to be moved into the user guide.

I suspect some of these changes may be contentious, I am happy to adjust this PR at reviewer discretion.

## How is this patch tested?

Locally + on databricks

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [x] `area/regression`: Sklearn regression pipeline example
- [ ] `area/build`: Build and test infrastructure for MLflow pipeline example

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: pipeline.yaml [hunk 1]
```diff
@@ -1,88 +1,31 @@
...
-# variables such as {{ INGEST_DATA_LOCATION }}.
-
-# Variables must be dereferenced in a profile YAML file, located under `profiles/`.
-# See `profiles/local.yaml` for example usage. One may switch among profiles quickly by
-# providing a profile name such as `local` in the Pipeline object constructor:
-# `p = Pipeline(profile="local")`
-#
-# NOTE: YAML does not support tabs for indentation. Please use spaces and ensure that all YAML
-# files are properly formatted.
-
```
Comment (jinzhang21) [inline pipeline.yaml:12]: Let's keep these comments since they are still useful to help understand the configs and caveats.

File: pipeline.yaml [hunk 2]
```diff
@@ -1,88 +1,31 @@
...
 template: "regression/v1"
-# Specifies the dataset to use for model development
 data:
-  # Dataset locations on the local filesystem are supported, as well as HTTP(S) URLs and
-  # any other remote locations resolvable by MLflow, such as those listed in
-  # https://mlflow.org/docs/latest/tracking.html#artifact-stores
```
Comment (jinzhang21) [inline pipeline.yaml:18]: Keep L16 - L18

File: pipeline.yaml [hunk 3]
```diff
@@ -1,88 +1,31 @@
...
-  custom_loader_method: steps.ingest.load_file_as_dataframe
-  # If the `spark_sql` `format` is specified, the `sql` entry is used to specify a SparkSQL
-  # statement that identifies the dataset to use
+  format: {{INGEST_DATA_FORMAT}}
   sql: SELECT * FROM delta.`{{INGEST_DATA_LOCATION}}`
-  # If the `delta` `format` is specified, you can also configure the Delta table `version` to read
-  # or the `timestamp` at which to read data
-  # version: 2
-  # timestamp: 2022-06-01T00:00:00.000Z
-# specify the dataset to use for batch scoring.  All params serve the same function as in `data`
```
Comment (jinzhang21) [inline pipeline.yaml:33]: Keep L33

File: pipeline.yaml [hunk 4]
```diff
@@ -1,88 +1,31 @@
...
 data_scoring:
   location: {{INGEST_SCORING_DATA_LOCATION}}
-  format: {{INGEST_SCORING_DATA_FORMAT|default('parquet')}}
-  custom_loader_method: steps.ingest.load_file_as_dataframe
+  format: {{INGEST_SCORING_DATA_FORMAT}}
   sql: SELECT * FROM delta.`{{INGEST_SCORING_DATA_LOCATION}}`
-# Specifies the name of the column containing targets / labels for model training and evaluation
```
Comment (jinzhang21) [inline pipeline.yaml:39]: Keep L39

File: pipeline.yaml [hunk 5]
```diff
@@ -1,88 +1,31 @@
...
 target_col: "fare_amount"
 steps:
   split:
-    # Train/validation/test split ratios
-    split_ratios: {{SPLIT_RATIOS|default([0.75, 0.125, 0.125])}}
-    # Specifies the method to use to perform additional processing and cleaning on split datasets
+    split_ratios: {{SPLIT_RATIOS}}
```
Comment (jinzhang21) [inline pipeline.yaml:13]: I think we can simplify this line by removing SPLIT_RATIOS variable from profiles. Just hard code the values here.

File: pipeline.yaml [hunk 6]
```diff
@@ -1,88 +1,31 @@
...
     post_split_method: steps.split.process_splits
   transform:
-    # Specifies the method that defines the data transformations to apply during model inference
```
Comment (jinzhang21) [inline pipeline.yaml:48]: Keep L48

File: pipeline.yaml [hunk 7]
```diff
@@ -1,88 +1,31 @@
...
     transformer_method: steps.transform.transformer_fn
   train:
     using: estimator_spec
-    # Specifies the method that defines the estimator type and parameters to use for model training
```
Comment (jinzhang21) [inline pipeline.yaml:52]: Keep L52

File: pipeline.yaml [hunk 8]
```diff
@@ -1,88 +1,31 @@
...
     estimator_method: steps.train.estimator_fn
   evaluate:
-    # Sets performance thresholds that a trained model must meet in order to be eligible for
-    # registration to the MLflow Model Registry
```
Comment (jinzhang21) [inline pipeline.yaml:56]: Keep L55 - L56

File: pipeline.yaml [hunk 9]
```diff
@@ -1,88 +1,31 @@
...
     validation_criteria:
       - metric: root_mean_squared_error
         threshold: 10
-      - metric: mean_absolute_error
-        threshold: 50
-      - metric: weighted_mean_squared_error
-        threshold: 20
   register:
-    # Specifies the name of the Registered Model to use when registering a trained model to
-    # the MLflow Model Registry
```
Comment (jinzhang21) [inline pipeline.yaml:66]: Keep L65 - L66

File: pipeline.yaml [hunk 10]
```diff
@@ -1,88 +1,31 @@
...
     model_name: "taxi_fare_regressor"
-    # Indicates whether or not a model that fails to meet performance thresholds should still
-    # be registered to the MLflow Model Registry
     allow_non_validated_model: false
```
Comment (jinzhang21) [inline pipeline.yaml:26]: Can remove L70. `allow_non_validated_model` is not strictly required (default to False)

File: pipeline.yaml [hunk 11]
```diff
@@ -1,88 +1,31 @@
...
   predict:
-    # Specifically define the model URI to use in batch scoring here or use the latest model
-    # registered from the training DAG
-    # model_uri: "models/model.pkl"
-    # Specify the output path of the scored data from predict
     output_location: {{SCORED_OUTPUT_DATA_LOCATION}}
-    # Specify the output format of the scored data from predict
```
Comment (jinzhang21) [inline pipeline.yaml:77]: Keep the comments between L72 and L77.

File: steps/custom_metrics.py [hunk 12]
```diff
@@ -1,49 +1 @@
...
-             dictionaries with multiple metric entries as well.
-    """
-    return {
-        "weighted_mean_squared_error": mean_squared_error(
-            eval_df["prediction"],
-            eval_df["target"],
-            sample_weight=1 / eval_df["prediction"].values,
-        )
-    }
+# This file must be present for the MLP regression pipeline to run successfully.
```
Comment (jinzhang21) [inline steps/custom_metrics.py:1]: This is a little inconvenient. Let's follow up with removing such requirements to have empty files lying around.

File: steps/split.py [hunk 13]
```diff
@@ -1,37 +1,7 @@
...
-- ``process_splits``: Defines customizable logic for processing & cleaning the training, validation,
-  and test datasets produced by the data splitting procedure.
-"""
-
 from pandas import DataFrame
 
 
 def process_splits(
     train_df: DataFrame, validation_df: DataFrame, test_df: DataFrame
 ) -> (DataFrame, DataFrame, DataFrame):
```
Comment (jinzhang21) [inline steps/split.py:6]: We should follow up to remove requiring user to provide these functions. No need to change in this PR though.

**Review by jinzhang21 — APPROVED** [review]
LGTM. Thanks, @bbarnes52 !

Could you also check if we need or can we move
[pylintrc](https://github.com/bbarnes52/mlp-regression-template/blob/simplifytemplate/pylintrc) to `.github/` directory?

It'd also be nice to move `requirements.txt` to `requirements/` directory, and update the README correspondingly.

File: pipeline.yaml (hunk 1 above)
Comment (dbczumar) [reply pipeline.yaml:12]: @jinzhang21 Bilal asked us to simplify the YAML, which IMO includes reducing the number of lines dedicated to comments - perhaps he should weigh in here. I think that these configs should be documented elsewhere, e.g. in a README on the repo and on mlflow.org.

File: steps/custom_metrics.py (hunk 12 above)
Comment (dbczumar) [reply steps/custom_metrics.py:1]: I think we should keep this file as-is, same with ingest.py. I don't think it's a major source of complexity.

File: steps/split.py [hunk 14]
```diff
@@ -1,37 +1,7 @@
...
-        cleaned = df.dropna()
-        # Filter out invalid fare amounts and trip distance
-        cleaned = cleaned[
-            (cleaned["fare_amount"] > 0)
-            & (cleaned["trip_distance"] < 400)
-            & (cleaned["trip_distance"] > 0)
-            & (cleaned["fare_amount"] < 1000)
-        ]
-
-        return cleaned
```
Comment (dbczumar) [inline steps/split.py:35]: The quality of the model produced in the example will suffer if we don't do this cleaning. I think it's best to preserve this.

File: steps/transform.py [hunk 15]
```diff
@@ -1,64 +1,19 @@
...
-    """
-    Extend the input dataframe with pickup day of week and hour, and trip duration.
-    Drop the now-unneeded pickup datetime and dropoff datetime columns.
-    """
-    df["pickup_dow"] = df["tpep_pickup_datetime"].dt.dayofweek
-    df["pickup_hour"] = df["tpep_pickup_datetime"].dt.hour
-    trip_duration = (
-            df["tpep_dropoff_datetime"] - df["tpep_pickup_datetime"]
-    )
-    df["trip_duration"] = trip_duration.map(lambda x: x.total_seconds() / 60)
```
Comment (dbczumar) [inline steps/transform.py:24]: The quality of the model produced in the example will suffer if we don't do this feature engineering. I think it's best to preserve this.

**Pull request closed without merging**